[Consensus Voting] Double vote detection

durkmurder opened the pull request:
### Context 

https://github.com/dapperlabs/flow-go/issues/5781
https://github.com/dapperlabs/flow-go/issues/5740#issuecomment-904670491

With current iteration of consensus voting v2 we have implemented `VoteCollector` based on idea that it collects votes for one block. This makes hard to keep track of all cases of double voting.

Some great points were raised by by @huitseeker in https://github.com/dapperlabs/flow-go/issues/5740#issuecomment-904670491 and https://github.com/onflow/flow-go/pull/1140#pullrequestreview-736454870. 

I would like to propose to change our design to address vote collecting based on views instead of block proposals. Addressing by views makes easy to track double voting cases and makes things very clear and logical. Pruning and indexing becomes a lot easier compared to [proposal](https://github.com/onflow/flow-go/pull/1161). 

### What was changed

Previously we had one vote collector for a tuple `(view, blockID)` now we will have one vote collector for each `view` and votes with different `blockID` but same `view` will be handled by one vote collector. It's up to vote collector how to deal with case of double voting. 

When we have received block proposal we can clearly make decision about double voting. I am not so sure about case when we haven't received block proposal. Based on https://github.com/dapperlabs/flow-go/issues/5740#issuecomment-904670491 we can still do some checks but can't be sure unless we have received block proposal. 

### TODO: 

Once this PR receives enough approvals there are still some things to handle:

- ~~Add extra index for `CachingVoteCollector`, maybe inside of `PendingVotes` to track index votes by `blockID`.~~
- ~~Fix current tests~~
- Update logic of `CachingVoteCollector` with regards to double voting.

AlexHentschel commented on consensus/hotstuff/vote_collector.go:
```diff
@@ -41,21 +41,6 @@ func (ps VoteCollectorStatus) String() string {
-	// * ErrDifferentCollectorState if the VoteCollector's state is different than expectedCurrentStatus
-	// * ErrInvalidCollectorStateTransition if the given state transition is impossible
-	// * all other errors are unexpected and potential symptoms of internal bugs or state corruption (fatal)
-	ChangeProcessingStatus(expectedValue, newValue VoteCollectorStatus) error
```
Lets consider the following situation:
* In view 100, Alice is primary and proposes block `B1`.
* In addition, evil Eve also proposes a block, `B2`, even though it is not her turn. Nevertheless, Eve signs her proposal properly.

From the perspective of our node:
1. the node receives votes for `B1` and `B2` (without having seen either blocks)
2. block `B1` arrives
3. block `B2` arrives 
    * at this point, the HotStuff state machine would detect that `B2` is an invalid block:  <img width="885" alt="Screen Shot 2021-08-27 at 4 11 38 PM" src="https://user-images.githubusercontent.com/26322303/131197086-90a87c8f-077e-4564-920b-0a4d2e1f35e8.png"> (full state machine [here](https://drive.google.com/file/d/1la4jxyaEJJfip7NCWBM9YBTz6PK4-N9e/view?usp=sharing))
    * The EventHandler therefore would call https://github.com/onflow/flow-go/blob/d43a18fc733de90ddea8b81acaa2443d702624e8/consensus/hotstuff/vote_aggregator_v2.go#L48-L50

In the previous version, we would have had independent `VoteCollector`s for `B1` and `B2` and an explicit state transition function
* for `B1` we could transition the sate from `CachingVoteCollector` to `VerifyingVoteCollector`
* for `B2` we could transition the sate from `CachingVoteCollector` to `InvalidVoteCollector`

With the revisions proposed in this PR, I am not clear on how we would pass the information along the the `VoteCollector` that the votes for `B1` are fine to process, but the votes for `B2` should be slashed.

durkmurder commented:
I have updated already implemented logic, added some extra checks, extended documentation. This I believe is a good starting point for tracking double votes. Test coverage and more updates to this will be done in subsequent PRs.
Request another review @AlexHentschel @huitseeker @zhangchiqing

AlexHentschel commented on consensus/hotstuff/votecollector/statemachine_test.go:
```diff
@@ -0,0 +1,200 @@
+		secondVote := unittest.VoteFixture(func(vote *model.Vote) {
+			vote.View = firstVote.View
+			vote.SignerID = firstVote.SignerID
+		})
+
```
```suggestion
		}) // voted blockID is randomly sampled, i.e. it will be different from firstVote
```

zhangchiqing commented on consensus/hotstuff/votecollector/statemachine.go:
```diff
@@ -55,129 +73,174 @@ func NewStateMachine(base CollectionBase) *StateMachine {
-func (m *StateMachine) AddVote(vote *model.Vote) error {
+func (m *VoteCollector) AddVote(vote *model.Vote) error {
+	// Cache vote
+	err := m.votesCache.AddVote(vote)
```
Do we still cache the vote even if we have received the block?

AlexHentschel replied:
We want to keep _all_ received votes, so we can do forensics later (and scan for double voting across all votes for this view).

zhangchiqing commented on consensus/hotstuff/votecollector/vote_cache.go:
```diff
@@ -0,0 +1,133 @@
+// consumer later, when already lots of votes are cached. However, this should
+// be a rare occurrence (we except moderate performance overhead in this case).
+func (vc *VotesCache) RegisterVoteConsumer(consumer VoteConsumer) {
+	if consumer == nil {
+		return
+	}
```
why would this case exist?

AlexHentschel replied:
I originally put this in to avoid a panic when we try to feed a vote into the consumer later (in a different thread). 

But you are probably right: we generally don't check for nil values; so we probably should remove this check here too.

zhangchiqing commented on consensus/hotstuff/votecollector/combined_vote_processor.go:
```diff
@@ -0,0 +1,163 @@
+		log:   log,
+		block: proposal.Block,
+	}
+	processor.done.Store(false)
```
is this call necessary?

AlexHentschel replied:
Note that `processor.done` is _not_ a pointer. Instead, to increase data locality, I included it directly. 
But I guess we could still do: 
```golang
	processor := &CombinedVoteProcessor{
		log:   log,
		block: proposal.Block,
		done: *atomic.NewBool(false),
	}
```
(?)

AlexHentschel commented on consensus/hotstuff/votecollector/statemachine_test.go:
```diff
@@ -0,0 +1,200 @@
+		require.NoError(t, err)
+		processor.AssertCalled(t, "Process", vote)
+	})
+}
```
we could also test that unexpected Processor errors are passed up the call stack:
```suggestion
	s.T().Run("unexpected-VoteProcessor-errors-are-passed-up", func(t *testing.T) {
		unexpectedError := errors.New("some unexpected error")
		vote := unittest.VoteForBlockFixture(block, unittest.WithVoteView(s.view))
		processor.On("Process", vote).Return(unexpectedError).Once()
		err := s.collector.AddVote(vote)
		require.ErrorIs(t, err, unexpectedError)
	})
}
```

AlexHentschel commented on consensus/hotstuff/votecollector/statemachine_test.go:
```diff
@@ -0,0 +1,200 @@
+	time.Sleep(100 * time.Millisecond)
+
+	processor.AssertExpectations(s.T())
+}
```
another test we could include is:

```golang
// Test_VoteProcessorErrorPropagation verifies that unexpected errors from the `VoteProcessor`
// are propagated up the call stack (potentially wrapped), but are not replaced.
func (s *StateMachineTestSuite) Test_VoteProcessorErrorPropagation() {
	block := helper.MakeBlock(s.T(), helper.WithBlockView(s.view))
	processor := s.prepareMockedProcessor(block)

	unexpectedError := errors.New("some unexpected error")
	vote := unittest.VoteForBlockFixture(block, unittest.WithVoteView(s.view))
	processor.On("Process", vote).Return(unexpectedError).Once()
	err := s.collector.AddVote(vote)
	require.ErrorIs(s.T(), err, unexpectedError)
}
```

AlexHentschel commented on consensus/hotstuff/votecollector/vote_cache_test.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	// registering vote consumer has to fill consumedVotes using callback
+	cache.RegisterVoteConsumer(voteConsumer)
```
would suggest to test that all cached votes are fed into the consumer right away

```suggestion
	cache.RegisterVoteConsumer(voteConsumer)
	// all cached votes should be fed into the consumer right away
	require.Equal(t, expectedVotes, consumedVotes)
```

AlexHentschel reviewed (approved):
very nice. Thanks for the detailed documentation of all the tests. It makes it very easy to read the test.

durkmurder commented on consensus/hotstuff/vote_collectors.go:
```diff
@@ -1,25 +1,24 @@
+	GetOrCreateCollector(view uint64) (collector VoteCollector, created bool, err error)
 
-	// Prune the vote collectors whose view is below the given view
+	// PruneUpToView prunes the vote collectors whose view is below the given view.
+	// If `view` is smaller than the previous value, the previous value is kept
+	// and the sentinel mempool.DecreasingPruningHeightError is returned.
```
https://github.com/onflow/flow-go/issues/1267
